### PR TITLE
[z3/xxx] Conan v2 migration

### PR DESCRIPTION
**Pull request by MartinDelille**
Specify library name and version:  **z3/xxx**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/z3/all/conanfile.py [hunk 1]
```diff
@@ -66,64 +73,60 @@ def requirements(self):
...
 
     def build(self):
         apply_conandata_patches(self)
 
         if self.options.multiprecision == "mpir":
-            tools.save(os.path.join(self._build_subfolder, "gmp.h"), textwrap.dedent("""\
+            save(self, os.path.join(self.build_folder, "gmp.h"), textwrap.dedent("""\
                 #pragma once
                 #include <mpir.h>
                 """))
```
Comment (MartinDelille) [inline recipes/z3/all/conanfile.py:121]: Any idea why this patch doesn't work ?

Comment (SpaceIm) [reply recipes/z3/all/conanfile.py:121]: take a look to patches which are parts of this trick

File: recipes/z3/all/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -1,10 +1,9 @@
-cmake_minimum_required(VERSION 3.1)
-project(test_package)
+cmake_minimum_required(VERSION 3.8)
 
-include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
-conan_basic_setup(TARGETS)
+project(test_package CXX)
 
-find_package(Z3 REQUIRED CONFIG)
+find_package(z3 REQUIRED CONFIG)
```
Comment (SpaceIm) [inline recipes/z3/all/test_package/CMakeLists.txt:5]: ?

File: recipes/z3/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -1,10 +1,9 @@
...
 
 add_executable(${PROJECT_NAME} test_package.cpp)
-target_link_libraries(${PROJECT_NAME} PRIVATE z3::libz3)
+target_link_libraries(${PROJECT_NAME} PRIVATE z3::z3)
+target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_17)
```
Comment (SpaceIm) [inline recipes/z3/all/test_package/CMakeLists.txt:9]: ?

File: recipes/z3/all/conanfile.py (hunk 1 above)
Comment (MartinDelille) [reply recipes/z3/all/conanfile.py:121]: Ok I see. I guess the  CONAN_PKG::mpi and CONAN_INCLUDE_DIRS are not valid anymore. Could you point me what should I replace this with ?

File: recipes/z3/all/test_package/CMakeLists.txt (hunk 2 above)
Comment (MartinDelille) [reply recipes/z3/all/test_package/CMakeLists.txt]: Fixed here: a65320080

File: recipes/z3/all/conanfile.py [hunk 4]
```diff
@@ -66,64 +73,59 @@ def requirements(self):
...
+                    raise ConanInvalidConfiguration(f"{self.ref} requires C++{self._minimum_cpp_standard}, which your compiler does not support.")
             else:
                 self.output.info(
                     f"{self.name} requires C++17. Your compiler is unknown. Assuming it supports C++17.")
+        if is_msvc(self) and self.info.options.shared:
+            raise ConanInvalidConfiguration(f"{self.ref} can not be built as shared on Visual Studio and msvc.")
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version],
+                  destination=self.source_folder, strip_root=True)
```
Comment (ericriff) [inline recipes/z3/all/conanfile.py:92]: Isn't `destination=self.source_folder` the default?

Comment (MartinDelille) [reply recipes/z3/all/conanfile.py:92]: The default is `"."` (https://docs.conan.io/en/latest/reference/conanfile/tools/files/downloads.html#conan-tools-files-get)

File: recipes/z3/all/test_package/CMakeLists.txt (hunk 3 above)
Comment (MartinDelille) [reply recipes/z3/all/test_package/CMakeLists.txt]: reverted here: db32ef6c156759eb7d401f5d4645ecc838efb1f5

File: recipes/z3/all/conanfile.py [hunk 5]
```diff
@@ -56,6 +60,9 @@ def configure(self):
         if self.options.multiprecision == "internal":
             self.provides.append("gmp")
 
+    def layout(self):
+        cmake_layout(self)
+
     def requirements(self):
         self.output.info(
```
[the remark this answers is not in the document]
Comment (MartinDelille) [reply recipes/z3/all/conanfile.py:67]: why ?

File: recipes/z3/all/conanfile.py (hunk 4 above)
Comment (uilianries) [reply recipes/z3/all/conanfile.py:92]: Please, keep destination as an explicit parameter, it's easier for reviewers and for who reads the recipe for the first time.

File: recipes/z3/all/conanfile.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ericriff) [reply recipes/z3/all/conanfile.py:67]: The way github renders my comment is confusing. The odd part is the following line
```py
f"{self.name} will build using {self.options.multiprecision} multiprecision implementation.")
```

[the remark this answers is not in the document]
Comment (MartinDelille) [reply recipes/z3/all/conanfile.py:67]: I didn't changed anything to this part. It was added here (56c249be5d) by @victorpaleologue

File: recipes/z3/all/conanfile.py (hunk 4 above)
Comment (MartinDelille) [reply recipes/z3/all/conanfile.py:92]: Can we resolve this ?

Comment (uilianries) [reply recipes/z3/all/conanfile.py:92]: Yes.

File: recipes/z3/all/conanfile.py (hunk 1 above)
Comment (victorpaleologue) [reply recipes/z3/all/conanfile.py:120]: How is it odd? Can I help in any way?

Comment (MartinDelille) [reply recipes/z3/all/conanfile.py:120]: Thanks @victorpaleologue ! I won't be able to go further with this PR but feel free to use my work if you want to go further with it!

**Pull request closed without merging**